_posts/2025-10-08-#33300.md

```diff
@@ -10,17 +10,66 @@ status: upcoming
+
+
+## Fun Exercise
+
+
+- If you get the `cmpctblock` harness running, revert PR [#33296](https://github.com/bitcoin/bitcoin/pull/33296) and see if you can reproduce the crash on this [`Assume`](https://github.com/bitcoin/bitcoin/blob/689a32197638e92995dd8eb071425715f5fdc3a4/src/net_processing.cpp#L3333) statement.
+
+- Yeah I know, fuzzing is cool.
 
 
 ## Questions
 
-1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. Did you review the PR? Concept ACK, approach ACK, tested ACK, or NACK? What was your review approach? Were you able to get the fuzz test running?
+
+2. Where in the codebase are the main compact block helpers and processing logic? Name some of the relevant classes and functions. (Hint: nothing like a little search for `NetMsgType::CMPCTBLOCK` to get you started)
+
+3. The fuzz test sends `SENDCMPCT` messages with `high_bandwidth` randomly set. What is the difference between high bandwidth and low bandwidth compact block relay? How many high bandwidth peers are allowed and does the fuzz harness test this limit?
+
+4. Compare `-testdatadir` and the new `-fuzzcopydatadir`. Why is the latter useful for performance, and why isn’t a fresh `TestingSetup` that mines blocks on each iteration acceptable here?
+
+5. Look at `create_block` in the harness. How many transactions do the generated blocks contain, and where do they come from? What compact block scenarios might be missed with only a few transactions in a block?
+
+6. In commit [ecba804](https://github.com/bitcoin/bitcoin/pull/33300/commits/ecba804b0d034bbb6e61510cd2161b2ae449ac8d) `FinalizeHeader` is moved to `util.h` so it can be used by any fuzz test. What does this function do and why doesn’t it stall the fuzzer?
+
+7. Commit [e2f9214](https://github.com/bitcoin/bitcoin/pull/33300/commits/e2f921458913bcbbe74115cdb2174b0ab31784f2) sorts `m_dirty_blockindex` by block hash instead of pointer address. What non-determinism does this fix?  The author [notes](https://github.com/bitcoin/bitcoin/pull/33300#issuecomment-3308381089) this slows production code for no production benefit. Why can't [`EnableFuzzDeterminism()`](https://github.com/bitcoin/bitcoin/blob/acc7f2a433b131597124ba0fbbe9952c4d36a872/src/util/check.h#L34) be used here? How do you think this non-determinism should be best handled (if not the way the PR currently does)?
+
+8. The harness adds pre-filled transactions to a compact block beyond the coinbase (which is all we do in production code for now). Walk through the index calculations for pre-filling transactions at indices 1 and 3 in a 4-transaction block. Do you think this code could be simplified or cleaned up, and if so, what do you suggest?
```

Comment:
Prefilled transactions come up quite regularly in this club, so I think this might be worth expanding on in the notes a bit. Also, BIP152 spells it without a hyphen, so I think consistently adopting that spelling here would be easier.